### PR TITLE
[fix](datetime) Fix the error in date calculation that includes constants

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

before

```
mysql> select hours_add('2023-03-30 22:23:45.23452',8);
+-------------------------------------+
| hours_add('2023-03-30 22:23:45', 8) |
+-------------------------------------+
| 2023-03-31 06:23:45                 |
+-------------------------------------+


mysql> select date_add('2023-03-30 22:23:45.23452',8);
+------------------------------------+
| date_add('2023-03-30 22:23:45', 8) |
+------------------------------------+
| 2023-04-07 22:23:45                |
+------------------------------------+

mysql [test]>select hours_add('2023-03-30 22:23:45.23452',8);
+-------------------------------------------+
| hours_add('2023-03-30 22:23:45.23452', 8) |
+-------------------------------------------+
| 2023-03-31 06:23:45.000234                |
+-------------------------------------------+
```

after

```
mysql [test]>select hours_add('2023-03-30 22:23:45.23452',8);
+-------------------------------------------+
| hours_add('2023-03-30 22:23:45.23452', 8) |
+-------------------------------------------+
| 2023-03-31 06:23:45.23452                 |
+-------------------------------------------+
1 row in set (0.01 sec)

mysql [test]>select date_add('2023-03-30 22:23:45.23452',8);
+------------------------------------------+
| date_add('2023-03-30 22:23:45.23452', 8) |
+------------------------------------------+
| 2023-04-07 22:23:45.23452                |
+------------------------------------------+
1 row in set (0.00 sec)

mysql [test]>set enable_nereids_planner=true;
Query OK, 0 rows affected (0.00 sec)

mysql [test]>set enable_fallback_to_original_planner=false;
Query OK, 0 rows affected (0.00 sec)

mysql [test]>select hours_add('2023-03-30 22:23:45.23452',8);
+-------------------------------------------+
| hours_add('2023-03-30 22:23:45.23452', 8) |
+-------------------------------------------+
| 2023-03-31 06:23:45.23452                 |
+-------------------------------------------+
1 row in set (0.03 sec)

mysql [test]>select date_add('2023-03-30 22:23:45.23452',8);
+------------------------------------------+
| days_add('2023-03-30 22:23:45.23452', 8) |
+------------------------------------------+
| 2023-04-07 22:23:45.23452                |
+------------------------------------------+
1 row in set (0.00 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Review by sohardforaname — APPROVED** [review]

**Comment by sohardforaname** [comment]
@morrySnow PTAL

**Comment by morrySnow** [comment]
we need to copy all your describe in Proposed changes to git commit msg, so please replace all screenshot in commit msg with text, thx

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/ComputeSignatureHelper.java [hunk 1]
```diff
@@ -123,7 +124,13 @@ private static FunctionSignature defaultDateTimeV2PrecisionPromotion(
                 continue;
             }
             if (finalType == null) {
-                finalType = DateTimeV2Type.forType(arguments.get(i).getDataType());
+                if (arguments.get(i) instanceof VarcharLiteral) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/ComputeSignatureHelper.java:127]: maybe u need use StringLikeLiteral

File: fe/fe-core/src/main/java/org/apache/doris/nereids/types/DateTimeV2Type.java [hunk 2]
```diff
@@ -78,6 +78,17 @@ public static DateTimeV2Type forType(DataType dataType) {
         return MAX;
     }
 
+    /**
+     * return proper type of datetimev2 for VarcharLiteral
+     * may be we need to check for validity?
+     */
+    public static DateTimeV2Type forTypeFromVarcharLiteral(String s) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/types/DateTimeV2Type.java:85]:
```suggestion
    public static DateTimeV2Type forExpression(Expression e) {
        if (e instanceof StringLikeLiteral) {
            ....
        } else {
            return forType(e.getDataType());
        }
```

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p1

File: fe/fe-core/src/main/java/org/apache/doris/nereids/types/DateTimeV2Type.java (hunk 2 above)
Comment (Mryange) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/types/DateTimeV2Type.java]: It has already been determined StringLikeLiteral, so it is unnecessary to check again.

**Comment by Mryange** [comment]
run p1

**Review by morrySnow — APPROVED** [review]

**Pull request merged**